sys/shell/commands/sc_gnrc_icmpv6_echo.c

```diff
@@ -0,0 +1,432 @@
+static int _finish(_ping_data_t *data);
+
+int _gnrc_icmpv6_ping(int argc, char **argv)
+{
+    _ping_data_t data = {
+        .netreg = GNRC_NETREG_ENTRY_INIT_PID(ICMPV6_ECHO_REP,
+                                                 sched_active_pid),
+        .count = DEFAULT_COUNT,
+        .tmin = UINT_MAX,
+        .datalen = DEFAULT_DATALEN,
+        .timeout = DEFAULT_TIMEOUT_USEC,
+        .interval = DEFAULT_INTERVAL_USEC,
+        .id = DEFAULT_ID,
+        .pattern = DEFAULT_ID,
+    };
+    int res;
+
+    if ((res = _configure(argc, argv, &data)) != 0) {
+        return res;
+    }
+    gnrc_netreg_register(GNRC_NETTYPE_ICMPV6, &data.netreg);
+    _pinger(&data);
+    do {
+        msg_t msg;
+
+        msg_receive(&msg);
+        switch (msg.type) {
+            case GNRC_NETAPI_MSG_TYPE_RCV: {
+                _handle_reply(&data, msg.content.ptr);
+                gnrc_pktbuf_release(msg.content.ptr);
+                break;
+            }
+            case _SEND_NEXT_PING:
+                _pinger(&data);
+                break;
+            case _PING_FINISH:
+                goto finish;
+            default:
+                /* requeue wrong packets */
+                msg_send(&msg, sched_active_pid);
+                break;
+        }
+    } while (data.num_recv < data.count);
+finish:
+    xtimer_remove(&data.sched_timer);
+    res = _finish(&data);
+    gnrc_netreg_unregister(GNRC_NETTYPE_ICMPV6, &data.netreg);
+    for (unsigned i = 0;
+         i < cib_avail((cib_t *)&sched_active_thread->msg_queue);
+         i++) {
+        msg_t msg;
+
+        /* remove all remaining messages (likely caused by duplicates) */
+        if ((msg_try_receive(&msg) > 0) &&
+            (msg.type == GNRC_NETAPI_MSG_TYPE_RCV) &&
+            (((gnrc_pktsnip_t *)msg.content.ptr)->type == GNRC_NETTYPE_ICMPV6)) {
+            gnrc_pktbuf_release(msg.content.ptr);
+        }
+        else {
+            /* requeue other packets */
+            msg_send(&msg, sched_active_pid);
+        }
+    }
+    return res;
+}
+
+static void _usage(char *cmdname)
+{
+    printf("%s [-c <count>] [-h] [-i <ms interval>] [-s <packetsize>]\n",
+           cmdname);
+    puts("     [-t hoplimit] [-W <ms timeout>] <host>[%<interface>]");
+    puts("     count: number of pings (default: 3)");
+    puts("     ms interval: wait interval milliseconds between sending "
+              "(default: 1000)");
+    puts("     packetsize: number of bytes in echo payload; must be >= 4 to "
+              "measure round trip time (default: 4)");
+    puts("     hoplimit: Set the IP time to life/hoplimit "
+              "(default: interface config)");
+    puts("     ms timeout: Time to wait for a resonse in milliseconds "
+              "(default: 1000). he option affects only timeout in absence "
+              "of any responses, otherwise wait for two RTTs");
+}
+
+static int _configure(int argc, char **argv, _ping_data_t *data)
```

Comment:
`ping6 -c 3` without hostname fails. I have some quite lightweight code for parsing command line arguments that I wrote several years ago. If there's interest I can bring it back to life. I think there's no point in everyone hacking together their own parser.

Reply:
Fixed.

Reply:
> I have some quite lightweight code for parsing command line arguments that I wrote several years ago. If there's interest I can bring it back to life. I think there's no point in everyone hacking together their own parser.

While I agree, but that opens a similar can of worms as the POSIX networking part, as some libc-variants already provide `getopt` and some don't. Since the `shell` is purely a debugging feature (and shell commands should also have as less dependencies as possible), I'd rather prefer every shell command to handle parsing for itself.

Reply:
getopt posix-only, and does not support long options (getopt_long is a GNU extension).

> I'd rather prefer every shell command to handle parsing for itself.

I would rephrase that as "I'd prefer every shell command to have it's own parsing bugs".

Reply:
> getopt posix-only, and does not support long options (getopt_long is a GNU extension).

At least my version of `newlib` supports it:

```
$ grep -R getopt /usr/arm-none-eabi/include/
/usr/arm-none-eabi/include/getopt.h:getopt.h - Read command line options
/usr/arm-none-eabi/include/getopt.h:The getopt() function parses the command line arguments.  Its arguments argc
/usr/arm-none-eabi/include/getopt.h:The getopt_long() function works like getopt() except that it also accepts
/usr/arm-none-eabi/include/getopt.h:The getopt() function returns the option character if the option was found
/usr/arm-none-eabi/include/getopt.h:The getopt_long() function's return value is described below.
/usr/arm-none-eabi/include/getopt.h:The function getopt_long_only() is identical to getopt_long(), except that a
/usr/arm-none-eabi/include/getopt.h:getopt() and friends to return EOF with optind != argc.
/usr/arm-none-eabi/include/getopt.h:This file and the accompanying getopt.c implementation file are hereby 
/usr/arm-none-eabi/include/getopt.h:    int *flag;			/* determines if getopt_long() returns a
/usr/arm-none-eabi/include/getopt.h:/* While getopt.h is a glibc extension, the following are newlib extensions.
/usr/arm-none-eabi/include/getopt.h: * They are optionally included via the __need_getopt_newlib flag.  */
/usr/arm-none-eabi/include/getopt.h:#ifdef __need_getopt_newlib
/usr/arm-none-eabi/include/getopt.h:     allocated variable of type struct getopt_data.  */
/usr/arm-none-eabi/include/getopt.h:  #define getopt_r		__getopt_r
/usr/arm-none-eabi/include/getopt.h:  #define getopt_long_r		__getopt_long_r
/usr/arm-none-eabi/include/getopt.h:  #define getopt_long_only_r	__getopt_long_only_r
/usr/arm-none-eabi/include/getopt.h:  /* The getopt_data structure is for reentrancy. Its members are similar to
/usr/arm-none-eabi/include/getopt.h:  typedef struct getopt_data
/usr/arm-none-eabi/include/getopt.h:  } getopt_data;
/usr/arm-none-eabi/include/getopt.h:#endif /* __need_getopt_newlib */
/usr/arm-none-eabi/include/getopt.h:  int _EXFUN (getopt,
/usr/arm-none-eabi/include/getopt.h:  int _EXFUN (getopt_long,
/usr/arm-none-eabi/include/getopt.h:  int _EXFUN (getopt_long_only,
/usr/arm-none-eabi/include/getopt.h:#ifdef __need_getopt_newlib
/usr/arm-none-eabi/include/getopt.h:  int _EXFUN (__getopt_r,
/usr/arm-none-eabi/include/getopt.h:	       struct getopt_data * __data));
/usr/arm-none-eabi/include/getopt.h:  int _EXFUN (__getopt_long_r,
/usr/arm-none-eabi/include/getopt.h:	       struct getopt_data * __data));
/usr/arm-none-eabi/include/getopt.h:  int _EXFUN (__getopt_long_only_r,
/usr/arm-none-eabi/include/getopt.h:	       struct getopt_data * __data));
/usr/arm-none-eabi/include/getopt.h:#endif /* __need_getopt_newlib */
/usr/arm-none-eabi/include/getopt.h:/* END OF FILE getopt.h */
/usr/arm-none-eabi/include/sys/unistd.h:# include <getopt.h>
/usr/arm-none-eabi/include/sys/unistd.h:extern char *optarg;			/* getopt(3) external variables */
/usr/arm-none-eabi/include/sys/unistd.h:int	 getopt(int, char * const [], const char *);
/usr/arm-none-eabi/include/sys/unistd.h:extern int optreset;			/* getopt(3) external variable */
```

Compare e.g. AVRlibc:

```
$ grep -R getopt /usr/lib/avr/include/
```